Support multitouch

igordmn opened the pull request:
1. Add multitouch handling for iOS. We receieve event with multi pointers from UIKitView, and redirect them to Compose (to ComposeScene)

2. Generate synthetic presses/releases, if we miss some press between 2 events. For example, on iOS simulator, multitouch sends evend without intermediate presses: `no presses -> event with 2 pressed fingers`. Compose doesn't work well this way, it needs events that sent sequentially (as in real life): `no presses -> event with 1 pressed fingers -> event with 2 pressed fingers`.

3. Add ImageViewer example, that shows multitouch with touch devices, and mouse scrolling with mouse

New API:
- ComposeScene.Pointer - represents a separate pointer (a single mouse, or multiple touches)
- ComposeScene.sendPointerEvent(pointers: List<Pointer>) - method that supports sending multiple pointer to Compose

Fixes https://github.com/JetBrains/compose-multiplatform/issues/2802

MatkovIvan commented on compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:
```diff
@@ -619,6 +661,70 @@ class ComposeScene internal constructor(
+        val position: Offset,
+
+        /**
+         * `true` if the pointer event is considered "pressed." For example, finger
+         *  touching the screen or a mouse button is pressed [pressed] would be `true`.
```
```suggestion
         * `true` if the pointer event is considered "pressed". For example,
         * a finger touches the screen or any mouse button is pressed.
```

MatkovIvan commented on compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:
```diff
@@ -619,6 +661,70 @@ class ComposeScene internal constructor(
+        val id: PointerId,
+
+        /**
+         * The [Offset] of the pointer.
```
relative to what?

igordmn replied:
The position inside `Pointer` can be relative to anything, depending on where it is used. When we use it as a parameter of `sendPointerEvent`, it is mentioned, relative to what it should be:
<img width="509" alt="image" src="https://user-images.githubusercontent.com/5963351/228788952-ee696e8c-51d0-4936-b94b-b0f83d48772e.png">

MatkovIvan commented on compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:
```diff
@@ -450,6 +451,7 @@ class ComposeScene internal constructor(
      * @param button Represents the index of a button which state changed in this event. It's null
      * when there was no change of the buttons state or when button is not applicable (e.g. touch event).
      */
+    @OptIn(ExperimentalComposeUiApi::class)
```
Is it considered as public API change?

igordmn replied:
No. `OptIn` is a compile-time annotation that supresses an error that we use experimental function inside.

It is fine to use experimental function inside, because they are implementation details, and are in the same module (if the experimental function in another module - we would have to support backward compatibility).

igordmn committed:
Support multitouch

1. Add multitouch handling for iOS. We receive event with multi pointers from UIKitView, and redirect them to Compose (to ComposeScene)

2. Generate synthetic presses/releases, if we miss some press between 2 events. For example, on iOS simulator, multitouch sends event without intermediate presses: `no presses -> event with 2 pressed fingers`. Compose doesn't work well this way, it needs events that sent sequentially (as in real life): `no presses -> event with 1 pressed fingers -> event with 2 pressed fingers`.

3. Add ImageViewer example, that shows multitouch with touch devices, and mouse scrolling with mouse

New API:
- ComposeScene.Pointer - represents a separate pointer (a single mouse, or multiple touches)
- ComposeScene.sendPointerEvent(pointers: List<Pointer>) - method that supports sending multiple pointer to Compose

Fixes https://github.com/JetBrains/compose-multiplatform/issues/2802
